pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -0,0 +1,220 @@
+	r.URL, err = newRegURL(string(text))
+	return err
+}
+
+// Registry is a generalized type representing a registry host and a registry
+// mirror, whereas the mirror cannot further be mirrored.
+type Registry struct {
+	URL      RegURL     `toml:"url"`
+	Blocked  bool       `toml:"blocked"`
+	Insecure bool       `toml:"insecure"`
+	Search   bool       `toml:"unqualified-search"`
+	Prefix   string     `toml:"prefix"`
+	Mirrors  []Registry `toml:"mirror"`
+}
+
+// String returns host:port/path of the Registry's URL.
+func (r *Registry) String() string {
+	return r.URL.String()
+}
+
+// tomlConfig is the data type used to marshal and unmarshal the toml config.
+type tomlConfig struct {
+	Registries []Registry `toml:"registry"`
+}
+
+// getUnqualifiedSearchRegistries returns all registries that are configured
+// for unqualified image search (i.e., with Registry.Search == true).
+func getUnqualifiedSearchRegistries(registries []Registry) []Registry {
+	unqualified := []Registry{}
+	for _, reg := range registries {
+		if reg.Search {
+			unqualified = append(unqualified, reg)
+		}
+	}
+	return unqualified
+}
+
+// FindRegistries returns all registries matching the specified image reference.
+// In case of an unqualified image reference, all registries that are configured
+// for unqualified image search are returned.  In case of a fully qualified
+// image reference, the registry with the longest prefix is returned.  An empty
+// array is returned in case no matching registry is found.
+func FindRegistries(imgRef reference.Reference, registries []Registry) ([]Registry, error) {
+	domain := reference.Domain(imgRef.(reference.Named))
+	// if no domain is specified, return a list of registries that enabled
+	// unqualified image search
+	if domain == "" {
+		regs := getUnqualifiedSearchRegistries(registries)
+		return regs, nil
+	}
+
+	regs := []Registry{}
+	reg := Registry{}
+	prefixLen := 0
+	for _, r := range registries {
+		if strings.HasPrefix(imgRef.String(), reg.Prefix) {
+			length := len(reg.Prefix)
+			if length > prefixLen {
+				reg = r
+				prefixLen = length
+			}
+		}
+	}
+	if prefixLen != 0 {
+		regs = append(regs, reg)
+	}
+	return regs, nil
+}
+
+// checkRegistries performs several sanity checks on the specified registries.
+// An error is returned in case a mirror is further mirrored.  Warnings are
+// printed if the a mirror's security level doesn't match the one of the
+// corresponding registry, and if a mirror has a specified Prefix, which has no
+// effect.
+func checkRegistries(registries []Registry) error {
+	for _, reg := range registries {
+		// default to the URL if Prefix isn't specified
+		if reg.Prefix == "" {
+			reg.Prefix = reg.String()
+		}
+		for _, mir := range reg.Mirrors {
+			if len(mir.Mirrors) > 0 {
+				return fmt.Errorf("mirror '%s' of registry '%s' cannot be mirrored", mir.String(), reg.String())
+			}
+			if mir.Prefix != "" {
+				logrus.Warnf("specified prefix for mirror '%s' has no effect", mir.String())
```

Comment:
It would be cleaner to add a separate `Mirror` type than to try to handle these inconsistencies in behavior after the fact. It really isn’t a recursive structure, so let’s not declare it like one.

Already this consistency check is costing about as many lines as reusing the `Registry` type has saved.

Reply:
:+1: